### PR TITLE
add upgrade tests to lifecycle suite

### PR DESCRIPTION
**Pull request by phoracek**
Test upgrades from previous releases to the current one.

### CODE REVIEW

File: test/e2e/lifecycle/upgrade_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+				CheckConfigCondition(ConditionAvailable, ConditionTrue, 10*time.Minute, CheckDoNotRepeat)
+				CheckReleaseUsesExpectedContainerImages(oldRelease)
+				CheckConfigVersions(oldRelease.Version, oldRelease.Version, oldRelease.Version, CheckImmediately, CheckDoNotRepeat)
+			})
+
+			Context("and it is upgraded to the latest release", func() {
+				BeforeEach(func() {
+					UninstallRelease(oldRelease)
+					InstallRelease(newRelease)
+					CheckOperatorIsReady(10 * time.Minute)
```
Comment (qinqon) [inline test/e2e/lifecycle/upgrade_test.go]: Constant for 10 * time.Minute, I see it all around

File: test/e2e/lifecycle/upgrade_test.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+			})
+
+			It("should transition through reported versions while being upgraded", func() {
+				// Upgrade the operator
+				UninstallRelease(oldRelease)
+				InstallRelease(newRelease)
+
+				// Check that operator and target versions will be set to the newer. Ignore observed version, since it
+				// might reach the target state immediately when no changes are needed between two releases
+				CheckConfigVersions(newRelease.Version, CheckIgnoreVersion, newRelease.Version, 10*time.Minute, CheckDoNotRepeat)
```
Comment (qinqon) [inline test/e2e/lifecycle/upgrade_test.go]: pass some variables with proper name to CheckConfigVersions
expectedObserverVersion
expectedTargetVersion
expectredOperatorVersion

File: test/e2e/lifecycle/upgrade_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+				// Wait until the operator finishes configuration
+				CheckConfigCondition(ConditionAvailable, ConditionTrue, 10*time.Minute, CheckDoNotRepeat)
+
+				// Validate that observed version turned to the newer
+				CheckConfigVersions(newRelease.Version, newRelease.Version, newRelease.Version, 10*time.Minute, CheckDoNotRepeat)
```
Comment (qinqon) [inline test/e2e/lifecycle/upgrade_test.go]: ditto

File: test/check/check.go [hunk 4]
```diff
@@ -125,6 +127,62 @@ func CheckOperatorIsReady(timeout time.Duration) {
...
+	listOptions := client.ListOptions{}
+	key := opv1alpha1.SchemeGroupVersion.Group + "/version"
+	listOptions.SetLabelSelector(fmt.Sprintf("%s,%s != %s", key, key, currentVersion))
+
+	deployments := appsv1.DeploymentList{}
+	err := framework.Global.Client.List(context.Background(), &listOptions, &deployments)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(deployments).To(BeEmpty(), "Found leftover objects from the previous operator version")
+
+	daemonSets := appsv1.DeploymentList{}
```
Comment (qinqon) [inline test/check/check.go]: DaemonSet

File: test/check/check.go [hunk 5]
```diff
@@ -125,6 +127,62 @@ func CheckOperatorIsReady(timeout time.Duration) {
 	}
 }
 
+func CheckForLeftoverObjects(currentVersion string) {
+	listOptions := client.ListOptions{}
+	key := opv1alpha1.SchemeGroupVersion.Group + "/version"
+	listOptions.SetLabelSelector(fmt.Sprintf("%s,%s != %s", key, key, currentVersion))
+
```
Comment (qinqon) [inline test/check/check.go]: You can iterate all this since the List() expectes a runtime.Object.

File: test/e2e/lifecycle/upgrade_test.go (hunk 1 above)
Comment (phoracek) [reply test/e2e/lifecycle/upgrade_test.go]: Done

File: test/check/check.go (hunk 4 above)
Comment (phoracek) [reply test/check/check.go]: Done

File: test/e2e/lifecycle/upgrade_test.go (hunk 2 above)
Comment (phoracek) [reply test/e2e/lifecycle/upgrade_test.go]: Cool IDEs would put the name there, mine doesn't neither

File: test/e2e/lifecycle/upgrade_test.go (hunk 3 above)
Comment (phoracek) [reply test/e2e/lifecycle/upgrade_test.go]: Done

File: test/check/check.go (hunk 5 above)
Comment (phoracek) [reply test/check/check.go]: test/check/check.go:133:17: cannot convert "github.com/kubevirt/cluster-network-addons-operator/vendor/k8s.io/api/apps/v1".DeploymentList literal (type "github.com/kubevirt/cluster-network-addons-operator/vendor
/k8s.io/api/apps/v1".DeploymentList) to type "github.com/kubevirt/cluster-network-addons-operator/vendor/k8s.io/apimachinery/pkg/runtime".Object:                                                                 
        "github.com/kubevirt/cluster-network-addons-operator/vendor/k8s.io/api/apps/v1".DeploymentList does not implement "github.com/kubevirt/cluster-network-addons-operator/vendor/k8s.io/apimachinery/pkg/runti
me".Object (DeepCopyObject method has pointer receiver)

Comment (phoracek) [reply test/check/check.go]: Yeah, but runtime.Object doesn't implement .Items()

**Comment by phoracek** [comment]
ci test please

ci please don't explode with testing system error

**Comment by phoracek** [comment]
ci test please

File: test/check/check.go (hunk 5 above)
Comment (qinqon) [reply test/check/check.go]: Is not that, since you are passing and interface you don't need to pass pointer, meaning just remove & from argument.

**Pull request merged**